sentry/src/main/java/io/sentry/Scopes.java

```diff
@@ -0,0 +1,1093 @@
+package io.sentry;
+
+import io.sentry.clientreport.DiscardReason;
+import io.sentry.hints.SessionEndHint;
+import io.sentry.hints.SessionStartHint;
+import io.sentry.metrics.LocalMetricsAggregator;
+import io.sentry.metrics.MetricsApi;
+import io.sentry.protocol.SentryId;
+import io.sentry.protocol.SentryTransaction;
+import io.sentry.protocol.User;
+import io.sentry.transport.RateLimiter;
+import io.sentry.util.ExceptionUtils;
+import io.sentry.util.HintUtils;
+import io.sentry.util.Objects;
+import io.sentry.util.Pair;
+import io.sentry.util.TracingUtils;
+import java.io.Closeable;
+import java.io.IOException;
+import java.lang.ref.WeakReference;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.WeakHashMap;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public final class Scopes implements IScopes, MetricsApi.IMetricsInterface {
+
+  private final @NotNull IScope scope;
+  private final @NotNull IScope isolationScope;
+  // TODO just for debugging
+  @SuppressWarnings("UnusedVariable")
+  private final @Nullable Scopes parentScopes;
+
+  private final @NotNull String creator;
+
+  // TODO should this be set on all scopes (global, isolation, current)?
+  private final @NotNull SentryOptions options;
+  private volatile boolean isEnabled;
+  private final @NotNull TracesSampler tracesSampler;
+
+  // TODO should this go on global scope?
+  private final @NotNull Map<Throwable, Pair<WeakReference<ISpan>, String>> throwableToSpan =
+      Collections.synchronizedMap(new WeakHashMap<>());
+  private final @NotNull TransactionPerformanceCollector transactionPerformanceCollector;
+  private final @NotNull MetricsApi metricsApi;
+
+  Scopes(
+      final @NotNull IScope scope,
+      final @NotNull IScope isolationScope,
+      final @NotNull SentryOptions options,
+      final @NotNull String creator) {
+    this(scope, isolationScope, null, options, creator);
+  }
+
+  private Scopes(
```

Comment:
Seeing `IScope` and `Scopes` side by side here could be confusing to our users, since both are public APIs. I'm wondering if we could use some backwards compatible `interface ScopeContext : IScope {}` to make it clear that `Scopes` is the forking mechanism / API and `ScopeContext` simply the data holder.

Reply:
`Scope` is part of https://develop.sentry.dev/sdk/unified-api/ so I don't think we should rename it.

Reply:
`Scopes` is in fact a plural of `Scope` as it combines multiple scopes together to allow scoping things like tags etc. to the users needs.